File: app/src/main/java/com/warchaser/musicplayer/tools/MusicList.java
package com.warchaser.musicplayer.tools;

import android.content.ContentResolver;
import android.content.ContentUris;
import android.database.Cursor;
import android.net.Uri;
import android.provider.MediaStore.Audio.Media;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by Wu on 2014/10/20.
 */
public class MusicList {

    public static List<MusicInfo> musicInfoList = new ArrayList<MusicInfo>();
    private static MusicList musicList;
    private static ContentResolver contentResolver;
    private Uri contentUri = Media.EXTERNAL_CONTENT_URI;
    private String order = "title COLLATE LOCALIZED";

    /**
     * The music which is playing.
     * */
    public static int iCurrentMusic;

    /**
     *The position of the music is playing.
     * */
    public static int iCurrentPosition;

    /**
     * The length of each music;
     * */
    public static int iCurrentMax;

    public static MyService.MyBinder mMyBinder;

    private String[] projection = {
            Media._ID,
            Media.TITLE,
            Media.DATA,
            Media.ALBUM,
            Media.ARTIST,
            Media.DURATION,
            Media.SIZE
    };

    private String sSortOrder = Media.DATA;

    public static MusicList instance(ContentResolver pContentResolver){
        if(null == musicList){
           contentResolver = pContentResolver;
            musicList = new MusicList();
        }
        return musicList;
    }

    private MusicList(){
        Cursor cursor = null;
        try{
            cursor = contentResolver.query(contentUri,null,null,null, order);
            if(null == cursor){
                return;
            }
            else if(!cursor.moveToFirst()){
                return;
            }
            else{
                int displayNameCol = cursor.getColumnIndex(Media.TITLE);
                int albumCol = cursor.getColumnIndex(Media.ALBUM);
                int albumIdCol = cursor.getColumnIndex(Media.ALBUM_ID);
                int idCol = cursor.getColumnIndex(Media._ID);
                int durationCol = cursor.getColumnIndex(Media.DURATION);
                int sizeCol = cursor.getColumnIndex(Media.SIZE);
                int artistCol = cursor.getColumnIndex(Media.ARTIST);
                int urlCol = cursor.getColumnIndex(Media.DATA);
                do{
                    String title = cursor.getString(displayNameCol);
                    String album = cursor.getString(albumCol);
                    long albumId = cursor.getLong(albumIdCol);
                    long id = cursor.getLong(idCol);
                    int duration = cursor.getInt(durationCol);
                    long size = cursor.getLong(sizeCol);
                    String artist = cursor.getString(artistCol);
                    String url = cursor.getString(urlCol);
                    MusicInfo musicInfo = new MusicInfo(id, title);
                    musicInfo.setAlbum(album);
                    musicInfo.setAlbumId(albumId);
                    musicInfo.setDuration(duration);
                    musicInfo.setSize(size);
                    musicInfo.setArtist(artist);
                    musicInfo.setUrl(url);

                    if(id > 0)
                    {
                        musicInfo.setUriWithCoverPic("content://media/external/audio/media/" + id + "/albumart");
                    }
                    else if(albumId > 0)
                    {
                        musicInfo.setUriWithCoverPic(ContentUris.withAppendedId(Uri.parse("content://media/external/audio/albumart"), albumId));
                    }
                    else
                    {
                        musicInfo.setUriWithCoverPic("");
                    }

//                    musicInfo.setPinyinInitial(StringHelper.getPingYin(title).substring(0,1).toUpperCase());
                    musicInfo.setPinyinInitial(StringHelper.getPingYin(title).substring(0,1));
                    if(musicInfo.getDuration() / 1000 > 60){
                        musicInfoList.add(musicInfo);
                    }

                }while(cursor.moveToNext());
                cursor.close();
            }
        }catch (Exception e){
            e.printStackTrace();
        }
    }

    public List<MusicInfo> getMusicList(){
        return musicInfoList;
    }

    public Uri getMusicUriById(long id){
        Uri uri = ContentUris.withAppendedId(contentUri, id);
        return uri;
    }

}

File: app/src/main/java/com/warchaser/musicplayer/tools/MusicList.kt
package com.warchaser.musicplayer.tools

import android.content.ContentResolver
import android.content.ContentUris
import android.database.Cursor
import android.net.Uri
import android.provider.MediaStore.Audio.Media
import java.io.File
import kotlin.collections.ArrayList

/**
 * Created by Wu on 2014/10/20.
 */
class MusicList private constructor(){

    companion object{
        private val MUSIC_INFO_LIST : ArrayList<MusicInfo> = ArrayList()
        private val CONTENT_URI : Uri = Media.EXTERNAL_CONTENT_URI
        private const val ORDER : String = "title COLLATE LOCALIZED"

        private lateinit var mContentResolver : ContentResolver

        /**
         * The music which is playing.
         */
        private var mCurrentMusic : Int = 0

        /**
         * The position of the music is playing.
         */
        private var mCurrentPosition : Int = 0

        /**
         * The length of each music;
         */
        private var mCurrentMax : Int = 0

        const val FIRST_POSITION : Int = 0
        const val FIRST_PROGRESS : Int = 0

        @JvmStatic
        fun init(contentResolver: ContentResolver){
            mContentResolver = contentResolver
            loadMusicList()
        }

        private fun loadMusicList(){
            var cursor : Cursor? = null
            try {
                var i : Int = 0
                cursor = mContentResolver.query(CONTENT_URI, null, null, null, ORDER)
                if(cursor == null){
                    return
                } else if(!cursor.moveToFirst()){
                    return
                } else {
                    val displayNameCol = cursor.getColumnIndex(Media.TITLE)
                    val albumCol = cursor.getColumnIndex(Media.ALBUM)
                    val albumIdCol = cursor.getColumnIndex(Media.ALBUM_ID)
                    val idCol = cursor.getColumnIndex(Media._ID)
                    val durationCol = cursor.getColumnIndex(Media.DURATION)
                    val sizeCol = cursor.getColumnIndex(Media.SIZE)
                    val artistCol = cursor.getColumnIndex(Media.ARTIST)
                    val urlCol = cursor.getColumnIndex(Media.DATA)
                    do {
                        val url : String = cursor.getString(urlCol)
                        if(!File(url).exists()){
                            continue
                        }

                        val title = cursor.getString(displayNameCol)
                        val album = cursor.getString(albumCol)
                        val albumId = cursor.getLong(albumIdCol)
                        val id = cursor.getLong(idCol)
                        val duration = cursor.getInt(durationCol)
                        val size = cursor.getLong(sizeCol)
                        val artist = cursor.getString(artistCol)

                        val musicInfo = MusicInfo(id, title)
                        musicInfo.album = album
                        musicInfo.albumId = albumId
                        musicInfo.duration = duration
                        musicInfo.size = size
                        musicInfo.artist = artist
                        musicInfo.url = url

                        when {
                            id > 0 -> {
                                musicInfo.uriWithCoverPic = "content://media/external/audio/media/$id/albumart"
                            }
                            albumId > 0 -> {
                                musicInfo.setUriWithCoverPic(getMediaStoreAlbumCoverUri(albumId))
                            }
                            else -> {
                                musicInfo.uriWithCoverPic = ""
                            }
                        }
//                    musicInfo.pinyinInitial = StringHelper.getPinYin(title).substring(0,1).toUpperCase(Locale.getDefault())
                        musicInfo.pinyinInitial = StringHelper.getPinYin(title).substring(0, 1)

                        if(musicInfo.duration / 1000 > 60){
                            MUSIC_INFO_LIST.add(musicInfo)
                        }

                        if(i <= CoverLoader.MAX_CACHE){
                            CoverLoader.instance.loadThumb(albumId)
                            CoverLoader.instance.loadBottomThumb(albumId)
                        }

                        i++

                    } while (cursor.moveToNext())

                    cursor.close()
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }

        @JvmStatic
        fun getMediaStoreAlbumCoverUri(albumId : Long) : Uri{
            val artworkUri = Uri.parse("content://media/external/audio/albumart")
            return ContentUris.withAppendedId(artworkUri, albumId)
        }

        @JvmStatic
        fun deleteSingleMusicFile(uri : String?) : Boolean{
            if(uri == null){
                return false
            }
            return File(uri).delete()
        }

        @JvmStatic
        fun isListNotEmpty() : Boolean{
            return MUSIC_INFO_LIST.isNotEmpty()
        }

        @JvmStatic
        fun getCurrentMusic() : MusicInfo{
            return MUSIC_INFO_LIST[mCurrentMusic]
        }

        @JvmStatic
        fun getMusicWithPosition(position : Int) : MusicInfo{
            return MUSIC_INFO_LIST[position]
        }

        @JvmStatic
        fun size() : Int{
            return MUSIC_INFO_LIST.size
        }

        @JvmStatic
        fun add(bean : MusicInfo){
            MUSIC_INFO_LIST.add(bean)
        }

        @JvmStatic
        fun remove(position: Int){
            if(MUSIC_INFO_LIST.isEmpty()){
                return
            }

            MUSIC_INFO_LIST.removeAt(position)
        }

        @Synchronized
        @JvmStatic
        fun setCurrentPosition(currentPosition : Int){
            mCurrentPosition = currentPosition
        }

        /**
         * 获取当前Music的播放进度
         * */
        @Synchronized
        @JvmStatic
        fun getCurrentPosition() : Int{
            return mCurrentPosition
        }

        @Synchronized
        @JvmStatic
        fun setCurrentMusic(currentMusic : Int){
            mCurrentMusic = currentMusic
        }

        /**
         * 获取当前Music的Index
         * */
        @Synchronized
        @JvmStatic
        fun getCurrentMusicInt() : Int{
            return mCurrentMusic
        }

        @Synchronized
        @JvmStatic
        fun setCurrentMusicMax(max : Int){
            mCurrentMax = max
        }

        /**
         * 获取当前Music最大进度
         * */
        @Synchronized
        @JvmStatic
        fun getCurrentMusicMax() : Int{
            return mCurrentMax
        }

        /**
         * 是否最后一首
         * */
        @JvmStatic
        fun isLastMusic() : Boolean{
            return getCurrentMusicInt() == size() - 1
        }

        /**
         * 获取下一首位置
         * */
        @JvmStatic
        fun getNextPosition() : Int{
            return getCurrentMusicInt() + 1
        }

        /**
         * 是否第一首
         * */
        @JvmStatic
        fun isFirstMusic() : Boolean{
            return getCurrentMusicInt() == 0
        }

        /**
         * 获取前一首位置
         * */
        @JvmStatic
        fun getPreviousPosition() : Int{
            return getCurrentMusicInt() - 1
        }

        /**
         * 获取最后一首位置
         * */
        @JvmStatic
        fun getLastPosition() : Int{
            return size() - 1
        }
    }

    fun getMusicList() : ArrayList<MusicInfo>{
        return MUSIC_INFO_LIST
    }

    fun getMusicUriById(id : Long) : Uri{
        return ContentUris.withAppendedId(CONTENT_URI, id)
    }

}